FullStorySegmentMiddleware/Classes/FullStoryMiddleware.m

```diff
@@ -124,6 +128,96 @@ - (SEGPayload *) getNewPayloadWithFSURL:(SEGContext * _Nonnull)context {
+    //TODO: Segment will crash and not allow props to have curcular dependency, but we should handle it here anyways
+    
+    NSMutableDictionary *props = [[NSMutableDictionary alloc] initWithCapacity:[properties count]];
+    for(id key in properties){
+        NSString *suffix = @"";
+
+        // properties should always be a NSDictionary, but still check to make sure the value is an object
+        // Check Type Encoding: https://developer.apple.com/library/archive/documentation/Cocoa/Conceptual/ObjCRuntimeGuide/Articles/ocrtTypeEncodings.html
+        const char* typeCode = @encode(typeof([props valueForKey:key]));
+
+        if([@"@" isEqualToString:@(typeCode)]){
+            NSObject *obj = [properties valueForKey:key];
+            
+            // if it's a NSDictionary: recurrsively get props, and we can not take underscore in key for nested objects, so we just remove them
+            // more info: https://help.fullstory.com/hc/en-us/articles/360020623234-FS-Recording-Client-API-Requirements
+            if([obj isKindOfClass:[NSDictionary class]]){
+                [props setValue:[self getSuffixedProps:[properties valueForKey:key]]
+                         forKey:[key stringByReplacingOccurrencesOfString:@"_" withString:@""]];
+            }else if ([obj isKindOfClass:[NSArray class]]){
+                // array of dicts? array of strings/ints/numbers? yuck!
+                NSMutableArray *arr = (NSMutableArray *) obj;
+                NSDictionary *dict = [self getObjectFromArrayObject:arr withKey:key];
+                [props addEntriesFromDictionary:dict];
+            }else{
+                suffix = [self getSuffixStringFromSimpleObject:obj];
+                [props setValue:[properties valueForKey:key] forKey:[key stringByAppendingString:suffix]];
+            }
+        }else{
+            #ifdef DEBUG
+                NSAssert(FALSE, @"key `%@` is not an object can't be serialized for FS custom event.", key);
+            #else
+                NSLog(@"key `%@` is not an object can't be serialized for FS custom event.", key);
+                // if prod, then don't send it
+            #endif
+        }
+    }
+    return props;
+}
+
+- (NSString *) getSuffixStringFromSimpleObject: (NSObject *) obj{
+    NSString * suffix;
+    // defualt to string
+    if([obj isKindOfClass:[NSNumber class]]){
+        // defaut to real
+        suffix = @"_real";
+        NSNumber *n = (NSNumber *) obj;
+        const char *type = n.objCType;
+        if([@"i" isEqualToString:@(type)]) suffix = @"_int";
+        // bool type number doesn't get encoded into 'B' but check anyway
+        else if ([@"B" isEqualToString:@(type)])suffix = @"_bool";
+    }else if ([obj isKindOfClass:[NSDate class]]){
+        suffix = @"_date";
+    }else if([obj isKindOfClass:[NSString class]]){
+        suffix = @"_str";
+    }
+    return suffix;
+}
+
+- (NSDictionary *) getObjectFromArrayObject: (NSArray *) arr withKey:(NSString *) key{
```

Comment:
nit: consistency in spacing -- I woudl probably write this `- (NSDictionary *) getObjectFromArrayObject:(NSArray *) arr withKey:(NSString *) key {`

Reply:
I look through our code in mn, seems like it has the spacing tightened like this: 
- (NSDictionary *) getObjectFromArrayObject:(NSArray *)arr withKey:(NSString *)key {...